### PR TITLE
Spherical ECFP

### PR DESCRIPTION
**Pull request by rlatawiec**
Added support for spherical ECFP and ECFP on subsets of atoms.

### CODE REVIEW

File: oddt/fingerprints.py [hunk 1]
```diff
@@ -324,34 +330,46 @@ def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
...
-                for neighbor in mol.Mol.GetAtomWithIdx(atom_idx).GetNeighbors():
-                    if neighbor.GetAtomicNum() == 1:
-                        continue
-                    n_idx = neighbor.GetIdx()
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
+        if mode == 'spherical':
+            neighbors = kd_tree.query_ball_point(mol.coords[idx], r=r*2)
+            for neighbor in neighbors:
+                atom = mol.OBMol.GetAtom(neighbor + 1)
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: `neighbor` is `Atom` object. `Atom` has both `atomicnum`  and `idx0` property. Querying for OBAtom is redundant and toolkit dependent.

Comment (mwojcikowski) [inline oddt/fingerprints.py]: Forget the neighbor part. I was wrong, this is still toolkit dependent which is wrong.

File: oddt/fingerprints.py [hunk 2]
```diff
@@ -323,35 +329,49 @@ def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
...
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
-            else:
-                for neighbor in mol.Mol.GetAtomWithIdx(atom_idx).GetNeighbors():
-                    if neighbor.GetAtomicNum() == 1:
-                        continue
-                    n_idx = neighbor.GetIdx()
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
+        if mode == 'spherical':
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: Indentation is clearly wrong.

File: oddt/fingerprints.py [hunk 3]
```diff
@@ -323,34 +329,49 @@ def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
...
-                    if neighbor.GetAtomicNum() == 1:
-                        continue
-                    n_idx = neighbor.GetIdx()
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
-        atom_env.append(atom_env[r - 1] + tmp)
+    if mode == 'spherical':
+        for r in range(1, depth + 1):
+            tmp = []
+            neighbors = kd_tree.query_ball_point(mol.coords[idx], r=r*2)
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: Check if it is faster from doing one full cdist() outside for loop.

File: oddt/fingerprints.py [hunk 4]
```diff
@@ -281,7 +282,8 @@ def _ECFP_atom_repr(mol, idx, use_pharm_features=False):
 
 
 def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
-                    atom_repr_dict=None, return_smiles=False, return_env=False):
+                    atom_repr_dict=None, return_smiles=False, return_env=False,
+                    mode='bond', kd_tree=None):
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: `mode` has no docstring.

File: oddt/fingerprints.py [hunk 5]
```diff
@@ -391,7 +412,7 @@ def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
 
 
 def ECFP(mol, depth=2, size=4096, count_bits=True, sparse=True,
-         use_pharm_features=False):
+         use_pharm_features=False, mode='bond', subset=None):
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: Docstrings missing.

File: oddt/fingerprints.py [hunk 6]
```diff
@@ -439,26 +472,41 @@ def ECFP(mol, depth=2, size=4096, count_bits=True, sparse=True,
...
+        mol_hashed = np.array(mol_hashed)
+        ecfp = mol_hashed.transpose(1, 0, 2)[0].astype(np.int64)
+        ecfp = np.sort(np.concatenate(ecfp))
+        idxs = np.argsort(ecfp)
+        sorted_smiles = np.concatenate(mol_hashed.transpose(1, 0, 2)[1])[idxs]
+    else:
+        ecfp = np.array(sorted(chain(*mol_hashed)))
+        print(mol_hashed)
     # folding
-    mol_hashed = fold(mol_hashed, size)
```
Comment (mwojcikowski) [inline oddt/fingerprints.py]: Please do not rename variables.

**Comment by rlatawiec** [comment]
ECFP performance comparison (master to experimental-fingerprints, bond-wise, no smiles).
Protein 2YMD used and its corresponding ligand.

depth=4, ligand:
- Master, 1.82 ms
- experimental, 1.8 ms

depth=4, protein:
- Master, 2.51 s
- experimental, 2.47 s

depth=2, ligand:
- Master, 867 µs
- experimental, 860 µs

depth=2, protein:
- Master, 1.14 s
- experimental, 1.12 s

**Pull request merged**